crates/telio-starcast/src/starcast_peer.rs

```diff
@@ -0,0 +1,398 @@
+use x25519_dalek::{PublicKey as PublicKeyDalek, StaticSecret};
+
+/// Constant for maximum packet size.
+const MAX_PACKET: usize = 2048;
+
+#[derive(Debug, thiserror::Error)]
+/// Custom `UdpProxy` error
+pub enum Error {
+    /// `StarcastPeer` Not Configured Error
+    #[error("Starcast peer is not configured")]
+    NotConfigured,
+    /// `StarcastPeer` `IoError`
+    #[error("Starcast peer IoError")]
+    IoError(#[from] std::io::Error),
+    /// Task execution failed
+    #[error(transparent)]
+    Task(#[from] telio_task::ExecError),
+    /// IP network failed
+    #[error("Failed to parse IP network")]
+    IpNetworkError(#[from] ipnetwork::IpNetworkError),
+    /// Send failed
+    #[error(transparent)]
+    Send(#[from] SendTimeoutError<(PublicKey, DataMsg)>),
+    /// Tunnel creation failed
+    #[error("Tunnel creation failed {0}")]
+    TunnNewFail(&'static str),
+}
+
+/// The actual starcast peer containing the task which runs the wait loop for this whole component.
+pub struct StarcastPeer {
+    /// starcast virtual peer task for Telio runtime.
+    task: Task<State>,
+}
+
+/// Starcast virtual peer configuration.
+pub struct Config {
+    /// Public key of the peer with which the starcast virtual peer shall communicate.
+    public_key: PublicKey,
+    /// Port number through which the virtual starcast peer shall communicate with WireGuard.
+    wg_port: u16,
+}
+
+impl StarcastPeer {
+    /// Create and start a new unconfigured starcast virtual peer. Note that it must first be configured to be functional.
+    /// A port is assigned dynamically for it's socket and also a public key which may be obtained by the application
+    /// by using the `get_peer()` method.
+    ///
+    /// # Arguments
+    ///
+    /// * `channel` - Channel for communicating decapsulated packets to and from the multicaster.
+    ///
+    /// # Returns
+    ///
+    /// A new starcast virtual peer.
+    pub async fn start(channel: Chan<Vec<u8>>, allowed_ips: Vec<IpNetwork>) -> Result<Self, Error> {
+        // Port 0 means the OS will dynamically allocate port.
+        const TIMER_UPDATE_PERIOD: Duration = Duration::from_millis(25);
+        let socket = UdpSocket::bind(SocketAddr::from(([127, 0, 0, 1], 0))).await?;
+        let wg_timer_tick_interval = interval(TIMER_UPDATE_PERIOD);
+        let secret_key = SecretKey::gen();
+
+        Ok(StarcastPeer {
+            task: Task::start(State {
+                secret_key,
+                channel,
+                socket,
+                tunnel: None,
+                wg_sock_addr: None,
+                wg_timer_tick_interval,
+                packets_present_in_tunnel: false,
+                sending_buffer: vec![0u8; MAX_PACKET],
+                receiving_buffer: vec![0u8; MAX_PACKET],
+                allowed_ips,
+            }),
+        })
+    }
+
+    // TODO figure out the implications and possible side effects of reconfiguring the starcast peer mid operation.
```

Comment:
Seems like important thing to do. I'm afraid that leaving such todo in code will result in the issue not being investigated in the end.

Please create a separate jira task (under the same parent as 5124) to track this problem and change the `TODO` to something like `TODO($JIRA_ID)`.